### ISSUE TITLE
SIGSEGV server 3.12.1.1/3.13.0.8

### ISSUE THREAD

**Issue by Aloren** [aerospike/aerospike-server#213]
We are using Aerospike docker container 3.12.1.1/3.13.0.8 and Aerospike java client 4.0.1. 
Quite often our tests start failing with errors -3 -- Node is not available or 11 -- Connection refused. After looking into container's logs I've found the following:
```
Feb 20 2018 17:12:58 GMT: WARNING (info): (thr_info.c:6625) SINDEX CREATE : Index with the same index defn already exists or bin has already been indexed.
Feb 20 2018 17:12:59 GMT: INFO (scan): (scan.c:1324) starting udf-bg scan job 6422411969392092577 {TEST:(null)} priority 2
Feb 20 2018 17:12:59 GMT: INFO (scan): (scan.c:1384) finished udf-bg scan job 6422411969392092577 (0)
Feb 20 2018 17:12:59 GMT: INFO (info): (thr_info.c:1753) scan get-job 6422411969392092577 0
Feb 20 2018 17:13:02 GMT: INFO (scan): (scan.c:1324) starting udf-bg scan job 5135553752816729310 {TEST:(null)} priority 2
Feb 20 2018 17:13:02 GMT: WARNING (as): (signal.c:181) SIGSEGV received, aborting Aerospike Community Edition build 3.12.1.1 os ubuntu16.04
Feb 20 2018 17:13:02 GMT: WARNING (as): (signal.c:185) stacktrace: found 12 frames
Feb 20 2018 17:13:02 GMT: WARNING (as): (signal.c:185) stacktrace: frame 0: asd(as_sig_handle_segv+0x48) [0x48a124]
Feb 20 2018 17:13:02 GMT: WARNING (as): (signal.c:185) stacktrace: frame 1: /lib/x86_64-linux-gnu/libc.so.6(+0x354b0) [0x7f4ee0c8a4b0]
Feb 20 2018 17:13:02 GMT: WARNING (as): (signal.c:185) stacktrace: frame 2: asd(as_bin_particle_size+0x41) [0x4669b8]
Feb 20 2018 17:13:02 GMT: WARNING (as): (signal.c:185) stacktrace: frame 3: asd(as_storage_record_get_n_bytes_memory+0x36) [0x5033d9]
Feb 20 2018 17:13:02 GMT: WARNING (as): (signal.c:185) stacktrace: frame 4: asd(udf_storage_record_open+0x96) [0x4d3b75]
Feb 20 2018 17:13:02 GMT: WARNING (as): (signal.c:185) stacktrace: frame 5: asd(udf_master_apply+0x5a5) [0x51fdc5]
Feb 20 2018 17:13:02 GMT: WARNING (as): (signal.c:185) stacktrace: frame 6: asd(udf_master+0x52) [0x52007c]
Feb 20 2018 17:13:02 GMT: WARNING (as): (signal.c:185) stacktrace: frame 7: asd(as_udf_start+0x122) [0x520226]
Feb 20 2018 17:13:02 GMT: WARNING (as): (signal.c:185) stacktrace: frame 8: asd(as_tsvc_process_transaction+0x865) [0x4c8a3a]
Feb 20 2018 17:13:02 GMT: WARNING (as): (signal.c:185) stacktrace: frame 9: asd(run_tsvc+0x7c) [0x4c8cfa]
Feb 20 2018 17:13:02 GMT: WARNING (as): (signal.c:185) stacktrace: frame 10: /lib/x86_64-linux-gnu/libpthread.so.0(+0x76ba) [0x7f4ee1eec6ba]
Feb 20 2018 17:13:02 GMT: WARNING (as): (signal.c:185) stacktrace: frame 11: /lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7f4ee0d5b82d]
```
I think that the issue with SIGSEGV might be either because of `starting udf-bg scan job 6422411969392092577 {TEST:(null)}` or `Index with the same index defn already exists or bin has already been indexed`.

In our tests we are also using the following UDF:
```
function remove_expired(rec, expireTime)
    local recordTtl = record.ttl(rec)
    if (recordTtl > 0 and recordTtl < expireTime) then
        aerospike:remove(rec)
    end
end
```
The following configuration file is used for Aerospike:
```
# Aerospike database configuration file.

# This stanza must come first.
service {
	user root
	group root
	paxos-single-replica-limit 1 # Number of nodes where the replica count is automatically reduced to 1.
	pidfile /var/run/aerospike/asd.pid
	service-threads 4
	transaction-queues 4
	transaction-threads-per-queue 4
	proto-fd-max 15000
}

logging {
	# Send log messages to stdout
	console {
		context any info
	}
}

network {
	service {
		address any
		port 3000

		# Uncomment the following to set the `access-address` parameter to the
		# IP address of the Docker host. This will the allow the server to correctly
		# publish the address which applications and other nodes in the cluster to
		# use when addressing this node.
		# access-address <IPADDR>
	}

	heartbeat {

		# mesh is used for environments that do not support multicast
		mode mesh
		port 3002

		# use asinfo -v 'tip:host=<ADDR>;port=3002' to inform cluster of
		# other mesh nodes

		interval 150
		timeout 10
	}

	fabric {
		port 3001
	}

	info {
		port 3003
	}
}

namespace TEST {
    memory-size 100M
    storage-engine memory
}
```
Could you please give any clues what might be the issue and how can it be solved?

**Comment by gooding470** [comment]
Hi Nastya,

This crash may be an Aerospike server issue that we have already fixed.  (Possibly it's issue AER-5721 in release 3.13.0.6.  Also, there were a few similar issues that nobody had observed yet that we fixed around this time.)

It would be best if you could upgrade from 3.12.1.1, which is by now quite an old version.  If it is really an issue that was never fixed, of course we'd prefer to see it in our latest version.

If you have a production environment and need to do rolling upgrades, you should first upgrade to 3.13.0.9.  That by itself might solve this issue.  (You must go through 3.13, which is a 'jump' version, in order to continue on to our latest.  If you want to do this, please read the upgrade instructions carefully, and/or ask us for support.)  If you are just testing, and don't have a deployment that requires a rolling upgrade, you can just start a new cluster directly with our latest.

Today, we will release 3.16.0.1, which among other things has a change you inspired: now list and map element delete operations will return a 'record not found' error code if the record doesn't exist.  (If the record exists but the particular bin doesn't exist, it will return 'ok'.)  This behavior is consistent with our regular bin delete operation, so it was a very good change -- thanks for that!  So maybe you want to wait a day and try this release.

3.16.0.1 also has a new ordered-list type, and a policy to enable 'inverted' or complementary sets of elements to be returned for various list and map operations.

I understand the common choice to not take the very latest release, but in Aerospike's case new releases are almost always better than older ones -- our ratio of fixes to regressions is very high.

**Comment by Aloren** [comment]
Hi!
Thank you for the quick reply.
We are now in the process of setting up Aerospike in production, so I suppose we will use the latest version of Aerospike if possible. In tests it's a matter of seconds to use newer version, so I will check and post an update here.

> Today, we will release 3.16.0.1

That is great news! Thank you for the cooperation! :+1:

**Comment by gooding470** [comment]
Ok, that's excellent -- you won't have to go through a 3.13 upgrade!  Please do try 3.16.0.1 later today.  I forgot to mention that if you want to try the new list and map APIs you'll need the latest client too.  (The 'record not found' error change will of course show up with any client.)

**Comment by Aloren** [comment]
I've just checked against 3.15.1.4. And it still crashes :weary:
```
Feb 20 2018 21:38:10 GMT: WARNING (as): (signal.c:181) SIGSEGV received, aborting Aerospike Community Edition build 3.15.1.4 os ubuntu16.04
Feb 20 2018 21:38:10 GMT: WARNING (as): (signal.c:185) stacktrace: found 11 frames
Feb 20 2018 21:38:10 GMT: WARNING (as): (signal.c:185) stacktrace: frame 0: asd(as_sig_handle_segv+0x48) [0x483b14]
Feb 20 2018 21:38:10 GMT: WARNING (as): (signal.c:185) stacktrace: frame 1: /lib/x86_64-linux-gnu/libc.so.6(+0x354b0) [0x7fb11e6df4b0]
Feb 20 2018 21:38:10 GMT: WARNING (as): (signal.c:185) stacktrace: frame 2: asd(free+0x34) [0x51a2db]
Feb 20 2018 21:38:10 GMT: WARNING (as): (signal.c:185) stacktrace: frame 3: asd(as_bin_particle_destroy+0x46) [0x4629bc]
Feb 20 2018 21:38:10 GMT: WARNING (as): (signal.c:185) stacktrace: frame 4: asd(write_master_dim+0x29d) [0x517c8c]
Feb 20 2018 21:38:10 GMT: WARNING (as): (signal.c:185) stacktrace: frame 5: asd(write_master+0x553) [0x5185d8]
Feb 20 2018 21:38:10 GMT: WARNING (as): (signal.c:185) stacktrace: frame 6: asd(as_write_start+0x11f) [0x518b71]
Feb 20 2018 21:38:10 GMT: WARNING (as): (signal.c:185) stacktrace: frame 7: asd(as_tsvc_process_transaction+0x90e) [0x4bf01a]
Feb 20 2018 21:38:10 GMT: WARNING (as): (signal.c:185) stacktrace: frame 8: asd(thr_demarshal+0xff7) [0x49d5d7]
Feb 20 2018 21:38:10 GMT: WARNING (as): (signal.c:185) stacktrace: frame 9: /lib/x86_64-linux-gnu/libpthread.so.0(+0x76ba) [0x7fb11f9426ba]
```

I'm running concurrent test with 10 threads which uses key ordered Map. After that I get that warning in Aerospike server and it stops. Are there any tools that can be used to verify what is actually causing such behavior? Can it be that 100mb storage is not enough?

**Comment by Aloren** [comment]
I've just checked that without these concurrent tests everything works perfectly, so I suppose that might be an issue with extensive load?

**Comment by gooding470** [comment]
Ok, this is big news.

We may have to hold up our 3.16.0.1 release for a bit to see if we can understand this.  Someone will soon look at the exact backtrace, but from the call stack, I notice a call to free memory.  It may be a double-free or prior overwrite of memory.  If you can reproduce this easily, would you mind running with service context configuration item 'debug-allocations' set to 'all'?  Just add in the service context of the config file:
...
service {
    ...
    debug-allocations all
    ...
}
...
If it's an overwrite or double free, instead of segv there will be an assertion, which will give us much better information.  In any case, if you can easily reproduce this, we'd love to know how, so we can attempt to fix it.  (There is a small chance it is already fixed in 3.16.0.1, but if so it would be a lucky accident.  We're not aware of any symptom looking like this as of 3.15.1.4.)

**Comment by Aloren** [comment]
Ok. thanks. Will check right now

**Comment by gooding470** [comment]
Yes, thanks for the test, we're trying to run it now.

And yes, there wouldn't be any debug logs from 'debug-allocations', it would simply assert instead of segv.  But your log shows it is still a segv.  And we got the detailed stack trace, which does show it crashes in our free call before we even get to check for overwrites or double frees.  So it's something else, and we're doing our best to reproduce it.

If we can't figure this out within the next few hours, we may go ahead and release 3.16.0.1 anyway, and when we eventually solve it, we will quickly release a hotfix.

**Comment by gooding470** [comment]
Our maps code expert (Shu) is currently looking at this in parallel with our Java client expert (Brian) who's trying to reproduce it with your test.

Shu claims he did fix something with similar symptoms very recently (that would already be in 3.16.0.1) and is checking to see whether somehow it wasn't fixed (and why not?) in 3.15.1.4.

Brian is running your test on 3.16.0.1 and can't reproduce, although he may be lacking some specifics.  He is about to try on 3.15.1.4.

**Comment by gooding470** [comment]
Ok, Shu says there *was* a bug in 3.15.1.4 where if putItems() was called with 'values' as an empty map, this exact symptom occurred.  It has been fixed already in 3.16.0.1.  If you think there's a chance that's what you were doing, that's likely the issue.

Brian is now trying on 3.15.1.4 and can't reproduce yet.  He'd like to know what 'values' your test is using.

**Comment by Aloren** [comment]
Values are random for each thread, each thread saves 5 items:
```
{36_token={expireTs=1519255554173, sender={firstName=firstName, lastName=lastName, snId=31, snType=2, id=31}, typeId=10, id=1519255554173, token=36_token}, 
17_token={expireTs=1519255554173, sender={firstName=firstName, lastName=lastName, snId=16, snType=2, id=16}, typeId=10, id=1519255554173, token=17_token}, 
51_token={expireTs=1519255554173, sender={firstName=firstName, lastName=lastName, snId=47, snType=2, id=47}, typeId=10, id=1519255554173, token=51_token}, 
43_token={expireTs=1519255554173, sender={firstName=firstName, lastName=lastName, snId=39, snType=2, id=39}, typeId=10, id=1519255554173, token=43_token}, 
30_token={expireTs=1519255554173, sender={firstName=firstName, lastName=lastName, snId=24, snType=2, id=24}, typeId=10, id=1519255554173, token=30_token}}
```
<img width="877" alt="screen shot 2018-02-21 at 1 30 00 am" src="https://user-images.githubusercontent.com/1207414/36454981-d68dfb30-16a6-11e8-9195-b44de21dc26e.png">

**Comment by Aloren** [comment]
I suppose that might be an issue with nested map...

**Comment by Aloren** [comment]
> Shu says there was a bug in 3.15.1.4 where if putItems() was called with 'values' as an empty map, this exact symptom occurred. It has been fixed already in 3.16.0.1.
Sorry, I found that nasty test that created document with empty values. Thank you guys! I will wait for 3.16.0.1. 
Have a nice day!

**Comment by gooding470** [comment]
Brian has now confirmed in 3.15.1.4 that adding an empty map (if the bin or record isn't there) does crash, with similar stack trace.  (Actually there are a few distinct ways the crashes can happen.)  But he can't reproduce this yet without empty map values.

If the map is never empty, maybe it's not the problem we already fixed.  (I'm not sure what you mean about nested maps, are you using them?)  We'll keep trying, but maybe it will be best if we just go ahead and release 3.16.0.1, so you can try that.  Then we'll know for sure whether it's a separate issue or not.

**Comment by gooding470** [comment]
Ah, ok.  We'll go ahead and release 3.16.0.1 as planned.  Please try it and let us know whether there's still a problem.  (You should be able to do anything crazy, and it shouldn't crash.)

**Comment by Aloren** [comment]
> You should be able to do anything crazy, and it shouldn't crash

Well, I wouldn't say that saving empty map is a crazy thing :stuck_out_tongue_winking_eye: But I'm really happy now that I caught this in tests.

> I'm not sure what you mean about nested maps, are you using them?
Yes, just for grouping data inside record. Is that a bad practice for Aerospike?

**Comment by gooding470** [comment]
You're right, saving an empty map is not even close to the craziest thing that has been done with Aerospike, but I just meant you should be able to do anything at all, even beyond that.  The server should never crash.

And nested maps are perfectly fine to use.  (Although our secondary indexes will only index on lists' and maps' top-level integer and string values, so there's some limitation there.)

**Comment by gooding470** [comment]
Ok Nastya, 3.16.0.1 is now available.  (Finally.)  You'll see AER-5803 in the release notes, concerning the changed error code when deleting elements from non-existent records or bins.  Please let us know if that seems to behave as you expect now.  We didn't include a release note for this issue -- the crash -- but we will add one very soon.  Please run your tests and let us know if it seems fixed as we think, or if anything bad still happens.

**Comment by Aloren** [comment]
Thank you. I will check as soon it gets available on Docker hub. :blush:

**Comment by gooding470** [comment]
Deployed to public repo:   https://hub.docker.com/r/aerospike/aerospike-server/~/dockerfile/
and official repo: https://hub.docker.com/_/aerospike/

The official repo takes a couple days to show, but the public repo should be available now.

The release note for this crash will be added to the website tonight, as AER-5806.

**Comment by kportertx** [comment]
@gooding470 Edited the public link you provided, github was converting the '~' to '%7E' which would then result in a 404 when accessed.

**Comment by Aloren** [comment]
Thank you guys!
Looks like it works fine. Going to update our environment now. :muscle:
